modules/navier_stokes/src/fvkernels/CNSFVHLLCBase.C

```diff
@@ -0,0 +1,119 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CNSFVHLLCBase.h"
+#include "NS.h"
+#include "HLLCUserObject.h"
+#include "SinglePhaseFluidProperties.h"
+
+namespace nms = NS;
+using MetaPhysicL::raw_value;
+
+std::vector<std::unordered_map<const FaceInfo *, std::array<ADReal, 3>>>
+    CNSFVHLLCBase::_fi_to_wave_speeds;
+
+InputParameters
+CNSFVHLLCBase::validParams()
+{
+  InputParameters params = FVFluxKernel::validParams();
+  params.addRequiredParam<UserObjectName>(nms::fluid, "Fluid userobject");
+  return params;
+}
+
+CNSFVHLLCBase::CNSFVHLLCBase(const InputParameters & params)
+  : FVFluxKernel(params),
+    _fluid(dynamic_cast<FEProblemBase *>(&_subproblem)
+               ->getUserObject<SinglePhaseFluidProperties>(nms::fluid)),
+    _specific_internal_energy_elem(getADMaterialProperty<Real>(nms::specific_internal_energy)),
+    _specific_internal_energy_neighbor(
+        getNeighborADMaterialProperty<Real>(nms::specific_internal_energy)),
+    _rho_et_elem(getADMaterialProperty<Real>(nms::total_energy_density)),
+    _rho_et_neighbor(getNeighborADMaterialProperty<Real>(nms::total_energy_density)),
+    _vel_elem(getADMaterialProperty<RealVectorValue>(nms::velocity)),
+    _vel_neighbor(getNeighborADMaterialProperty<RealVectorValue>(nms::velocity)),
+    _speed_elem(getADMaterialProperty<Real>(nms::speed)),
+    _speed_neighbor(getNeighborADMaterialProperty<Real>(nms::speed)),
+    _rho_elem(getADMaterialProperty<Real>(nms::density)),
+    _rho_neighbor(getNeighborADMaterialProperty<Real>(nms::density)),
+    _pressure_elem(getADMaterialProperty<Real>(nms::pressure)),
+    _pressure_neighbor(getNeighborADMaterialProperty<Real>(nms::pressure))
+{
+  if (_tid == 0)
+    _fi_to_wave_speeds.resize(libMesh::n_threads());
+}
+
+HLLCData
+CNSFVHLLCBase::hllcData() const
+{
+  return {_fluid,
+          _rho_elem[_qp],
+          _rho_neighbor[_qp],
+          _vel_elem[_qp],
+          _vel_neighbor[_qp],
+          _specific_internal_energy_elem[_qp],
+          _specific_internal_energy_neighbor[_qp]};
+}
+
+const std::array<ADReal, 3> &
+CNSFVHLLCBase::waveSpeed(const THREAD_ID tid,
+                         const FaceInfo & fi,
+                         const HLLCData & hllc_data,
+                         const ADRealVectorValue & normal)
+{
+  mooseAssert(tid < libMesh::n_threads(), tid << " must be less than " << libMesh::n_threads());
+
+  auto it = _fi_to_wave_speeds[tid].find(&fi);
+  if (it != _fi_to_wave_speeds[tid].end())
```

Comment:
did you check that it is faster to cache ^^

Reply:
No 😆